### PR TITLE
Create asset price controller for getting asset USD values

### PR DESCRIPTION
**Pull request by bbondy**
Resolves https://github.com/brave/brave-browser/issues/16509

This work tracks the creation of a new controller that talks with our internal server for getting USD values for assets.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet_ui/page/async/wallet_page_async_handler.ts [hunk 1]
```diff
@@ -64,4 +64,9 @@ handler.on(WalletPageActions.walletBackupComplete.getType(), async (store) => {
   await refreshWalletInfo(store)
 })
 
+// TODO(bbondy): Remove - Example usage:
```
Comment (bbondy) [inline components/brave_wallet_ui/page/async/wallet_page_async_handler.ts:67]: This is intentionally meant to be merged for now and removed later.

File: components/brave_wallet_ui/utils/mojo-time.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { MojoTime } from '../constants/types'
+
+/**
+ * Converts a mojo time to a JS time.
+ * @param {!mojoBase.mojom.Time} mojoTime
+ * @return {!Date}
+ */
+export function convertMojoTimeToJS (mojoTime: MojoTime) {
```
Comment (bbondy) [inline components/brave_wallet_ui/utils/mojo-time.ts:8]: Currently unused but will be used in a followup.

File: browser/ui/webui/brave_wallet/common_handler/wallet_handler.cc [hunk 3]
```diff
@@ -75,6 +76,66 @@ void WalletHandler::UnlockWallet(const std::string& password,
   std::move(callback).Run(result);
 }
 
+void WalletHandler::GetAssetPrice(const std::string& asset,
+                                  GetAssetPriceCallback callback) {
+  auto* profile = Profile::FromWebUI(web_ui_);
+  auto* asset_ratio_controller =
+      GetBraveWalletService(profile)->asset_ratio_controller();
+  asset_ratio_controller->GetPrice(
+      asset, base::BindOnce(&WalletHandler::OnGetPrice, base::Unretained(this),
```
Comment (bbondy) [inline browser/ui/webui/brave_wallet/common_handler/wallet_handler.cc:85]: I think this needs a WeakPtr, doing now.

File: components/brave_wallet/browser/asset_ratio_controller.cc [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+// static
+GURL AssetRatioController::GetPriceURL(const std::string& asset) {
+  std::string passthrough = base::StringPrintf(
+      "/api/v3/simple/price?ids=%s&vs_currencies=usd", asset.c_str());
+  passthrough = net::EscapeQueryParamValue(passthrough, false);
+  std::string spec = base::StringPrintf(
+      "%sv2/coingecko/passthrough?path=%s",
+      base_url_for_test_.is_empty() ? "https://bat-ratios.herokuapp.com/"
```
Comment (yrliou) [inline components/brave_wallet/browser/asset_ratio_controller.cc:74]: nit: define https://bat-ratios.herokuapp.com/ as a constant.

**Review by yrliou — APPROVED** [review]

File: components/brave_wallet/browser/asset_ratio_controller.cc (hunk 4 above)
Comment (bbondy) [reply components/brave_wallet/browser/asset_ratio_controller.cc]: Updated and pushed

**Pull request merged**